sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/specialized/BlobAsyncClientBase.java

```diff
@@ -767,29 +745,201 @@ private AsynchronousFileChannel downloadToFileResourceSupplier(String filePath)
+
+                    // Make the download call.
+                    return this.downloadWithResponse(chunkRange, downloadRetryOptions, realConditions,
+                        rangeGetContentMd5, null)
+                        .subscribeOn(Schedulers.elastic())
+                        .flatMap(response ->
+                            writeBodyToFile(response, file, chunkNum, finalParallelTransferOptions, progressLock,
+                                totalProgress));
+                })
+                // All the headers will be the same, so we just pick the last one.
+                .last();
+        });
     }
 
-    private List<BlobRange> sliceBlobRange(BlobRange blobRange, Integer blockSize) {
-        if (blockSize == null) {
-            blockSize = BLOB_DEFAULT_DOWNLOAD_BLOCK_SIZE;
+    private int calculateNumBlocks(long dataSize, long blockLength) {
+        // Can successfully cast to an int because MaxBlockSize is an int, which this expression must be less than.
+        int numBlocks = toIntExact(dataSize / blockLength);
+        // Include an extra block for trailing data.
+        if (dataSize % blockLength != 0) {
+            numBlocks++;
         }
-        long offset = blobRange.getOffset();
-        long length = blobRange.getCount();
-        List<BlobRange> chunks = new ArrayList<>();
-        for (long pos = offset; pos < offset + length; pos += blockSize) {
-            long count = blockSize;
-            if (pos + count > offset + length) {
-                count = offset + length - pos;
+        return numBlocks;
+    }
+
+    /*
+    Download the first chunk. Construct a Mono which will emit the total count for calculating the number of chunks,
+    access conditions containing the etag to lock on, and the response from downloading the first chunk.
+     */
+    private Mono<Tuple3<Long, BlobRequestConditions, BlobDownloadAsyncResponse>> getSetupMono(BlobRange r,
+        ParallelTransferOptions o, DownloadRetryOptions downloadRetryOptions,
+        BlobRequestConditions requestConditions, boolean rangeGetContentMd5, Context context) {
+        // We will scope our initial download to either be one chunk or the total size.
+        long initialChunkSize = r.getCount() != null && r.getCount() < o.getBlockSize()
+            ? r.getCount() : o.getBlockSize();
+
+        return this.downloadWithResponse(new BlobRange(r.getOffset(), initialChunkSize), downloadRetryOptions,
+            requestConditions, rangeGetContentMd5, context)
+            .subscribeOn(Schedulers.elastic())
+            .flatMap(response -> {
+                /*
+                Either the etag was set and it matches because the download succeeded, so this is a no-op, or there
+                was no etag, so we set it here. ETag locking is vital to ensure we download one, consistent view
+                of the file.
+                 */
+                BlobRequestConditions newConditions = setEtag(requestConditions, response.getHeaders()
```

Comment:
Could get the ETag through the deserializedHeaders on the response object instead of accessing the raw header map, Same goes for the other places the raw headers map is being used.

Reply:
The one place I can't do this is when I'm checking the range after catching a StorageException. That one just has a generic response, so I have to go into the raw headers.